README.md
### Emojitwo Emoticon Theme for Pidgin <img src="svg/1f44d-1f3fd.svg" width="32" height="32" />

<!--
![thumbs up](svg/1f44d-1f3fd.svg =24x24)
-->

The [Emojitwo](https://emojitwo.github.io/) emoticon theme made usable with
[Pidgin](http://pidgin.im/) instant messanger.

Emoji artwork is provided by Emojitwo, originally released as
[Emojione 2.2](https://www.emojione.com/) by [Ranks.com](http://www.ranks.com/)
with contributions from the Emojitwo community and is licensed under CC-BY 4.0.

This version/release is not endorsed by Ranks.com or the Emojitwo community.

### Licensing

#### EmojiOne Artwork

*  Applies to all PNG and SVG files as well as any adaptations made.
*  License: Creative Commons Attribution 4.0 International
*  Human Readable License: http://creativecommons.org/licenses/by/4.0/
*  Complete Legal Terms: http://creativecommons.org/licenses/by/4.0/legalcode

### Creating/Exporting Theme

#### Commands

If [Python](https://www.python.org/) is installed (requires version 3+), the
included ***release.py*** script can be run to export the theme for use with
Pidgin.

```
Usage:
	release.py [options]

Description
	Helper script for creating release.

Options:
	-h | --help             Show this usage information.
	-O | --overwrite        Existing PNG images will be overitten with new ones.
	-n | --dry-run          No action is taken.
	-c | --clean            Cleans the directory tree.
	-s | --size             Size(s) for exported PNG images in pixels (e.g.
							--size 24,32,64).
	--no-update-template    Theme template will not be regenerated (only works
							if template exists).
	--force-update-template Forces the template to be updated (only useful with
							--dry-run).
	--all-images			Don't ignore images marked for exclusion in template
							file.
							Also removes images found in release that have been
							marked for exclusion.
	-t | --tag              Create Git tag from INFO file (requires "git"
							command).
```

The exported files will be created in the ***release*** directory. By default,
PNG images will be exported in 16x16 & 24x24 pixels. To specify which pixel
sizes to export use the ***--size*** option. It accepts a comma-separated list
(e.g. *--size 16,24,32*).

Example:

```
./release.py --size 16,24,32
```

Existing PNG images will not be overwritten when the theme is exported. To force
overwriting existing images use the ***--overwrite*** option:

```
./release.py --overwrite
```

If you only want to overwrite/update specific images, the ***--overwrite***
accepts a comma-separated list value:

```
./release.py --overwrite 263a.png,2639.png
```

#### Template File

A template file named ***[template.txt](template.txt)*** controls how the theme
will be exported. This file is automatically generated by the ***release.py***
script using information found in the ***[INFO](INFO)*** file & will add any new
images found in the ***svg*** directory to the ***default*** group. Lines
beginning with "#!" denote that the image will *not* be included in the exported
theme.

```
[default]
# included images
263a.png
2639.png
# excluded images
#!1f600.png
#!1f601.png
```

To override this behavior, use the ***--all-images*** option:

```
./release.py --all-images
```


CHANGES.txt
Pre-release:
- copied SVG emojis from Emojitwo Github repo at commit 396be0e
- created helper scripts to convert images from SVG to PNG & create releases
- used images 1F60A & 1F618 to create 1F970 (Smiling Face With Hearts) emoji
- copied SVG for unicode character U+1F929 (Star-Struck) from Emojitwo Github repo at commit ba2e126
- changed 1F929 eyes to purple


release.py
#!/usr/bin/env python

# This script is licensed on Creative Commons Zero (CC0).
# See: https://creativecommons.org/publicdomain/zero/1.0/


import os, shutil, sys

from py			import fileio
from py			import info
from py			import pyIsCompat
from py.cl		import args
from py.md		import markdownToText
from py.paths	import appendPath
from py.paths	import dir_release
from py.paths	import dir_svg
from py.paths	import file_license
from py.paths	import file_readme
from py.paths	import root as dir_root
from py.paths	import template_file
from py.theme	import copyTemplate
from py.theme	import getImagesToRemove
from py.theme	import getReleaseImages
from py.util	import compress
from py.util	import convertToPNG
from py.util	import execute
from py.util	import getCommand
from template	import init as generateTemplate


try:
	py_compat, py_ver = pyIsCompat()
	if not py_compat:
		print('\nERROR:\tUsing Python version {}. Version 3 or greater required.'.format(py_ver))
		sys.exit(1)


	if args.contains('clean'):
		# use condition to prevent IDE from organizing import
		if True: import clean

		sys.exit(clean.init(args.getValue('clean', True)))

	if args.contains('tag'):
		cmd_git = getCommand('git')
		if not cmd_git:
			print('ERROR: git command not found')
			sys.exit(1)

		ver_main = info.getAttribute('version')
		try:
			ver_package = int(info.getAttribute('package_release'))
		except (TypeError, ValueError):
			print('\nWARNING: Could not get package version from INFO file, setting to "0".')
			ver_package = 0

		new_tag = 'v{}'.format(ver_main)
		if ver_package:
			new_tag = '{}-{}'.format(new_tag, ver_package)

		print('\nCreating tag {} from INFO file ...'.format(new_tag))
		execute('{} tag {}'.format(cmd_git, new_tag))

		sys.exit(0)

	live_run = not args.contains('dry-run')
	if not live_run:
		print('\nDry run: Not making any changes!\n')

	# --force-update-template is used to regenerate template when --dry-run is specified.
	if not live_run and args.contains('force-update-template'):
		generateTemplate()
	elif live_run:
		if not args.contains('no-update-template') or not os.path.isfile(template_file):
			generateTemplate()

		# create output directory
		os.makedirs(dir_release, exist_ok=True)

	# default sizes (overridden with --sizes option)
	sizes = ['16', '24']
	if args.contains('size'):
		sizes = args.getValue('size', True)
	else:
		sizes_copy = list(sizes[:-1])
		sizes_copy.append('and {}'.format(sizes[-1]))
		sizes_delim = ' '
		if len(sizes) > 2:
			sizes_delim = ', '
		print('\nCreating releases for image sizes {} pixels.'.format(sizes_delim.join(sizes_copy)))
		print('This can be changed by using the "--size" option.')

	for S in sizes:
		# check that all sizes are numerical values
		try:
			int(S)
		except ValueError:
			print('\nERROR: "{}" is not a valid numerical value for argument "size".'.format(S))
			sys.exit(1)

	# images to be included with release
	include_all = args.contains('all-images')
	release_images = getReleaseImages(all_images=include_all)
	img_count = len(release_images)

	# images that should be removed from release if found
	remove_images = ()
	if not include_all:
		remove_images = getImagesToRemove()

	print('{} images will be included in release.'.format(img_count))

	# images that will be overwritten if found in release (default is to not overwrite any, use --overwrite to set)
	overwrite_images = None
	if args.contains('overwrite-png'):
		overwrite_images = args.getValue('overwrite-png')
		if ',' in overwrite_images:
			overwrite_images = tuple(overwrite_images.split(','))

	# prepare README for inclusion in release
	readme_text = fileio.read(file_readme)
	# don't include export section
	if '### Creating/Exporting Theme' in readme_text:
		readme_text = readme_text.split('### Creating/Exporting Theme')[0]
	readme_text = markdownToText(readme_text)

	grp_count = len(sizes)
	grp_idx = 0
	for S in sizes:
		converted_count = 0
		replaced_count = 0
		removed_count = 0
		grp_idx += 1

		print('\nProcessing {}x{} images (group {}/{}) (Ctrl+C to cancel) ...'.format(S, S, grp_idx, grp_count))

		idx = 0
		size_dir = appendPath(dir_release, '{}/emojitwo'.format(S))

		# check for previously generated release
		if remove_images and os.path.isdir(size_dir):
			print('\nChecking images from previous release ...')
			# images found in target release directory
			png_existing = os.listdir(size_dir)

			# exclude non-PNG files
			for p_idx in reversed(range(len(png_existing))):
				if not png_existing[p_idx].lower().endswith('.png'):
					png_existing.pop(p_idx)

			if not png_existing:
				print('... none found.')

			# check if files should not be included in release
			for PNG in png_existing:
				if PNG in remove_images:
					sys.stdout.write('Removing existing image from release: {}            \r'.format(PNG))
					if live_run:
						PNG = appendPath(size_dir, PNG)
						os.remove(PNG)
					removed_count += 1

		for img_name in release_images:
			idx += 1

			source = appendPath(dir_svg, '{}.svg'.format(img_name))
			target = appendPath(size_dir, '{}.png'.format(img_name))

			os.makedirs(size_dir, exist_ok=True)

			img_exists = os.path.isfile(target)
			if img_exists:
				# check if image should be overwritten (default is no)
				overwrite = False
				if overwrite_images and type(overwrite_images) == tuple:
					overwrite = '{}.png'.format(img_name) in overwrite_images
				elif overwrite_images:
					overwrite = '*' == overwrite_images or '{}.png'.format(img_name) == overwrite_images

				if not overwrite:
					sys.stdout.write('Not updating PNG image ({}/{})            \r'.format(idx, img_count))
					continue

			if live_run:
				sys.stdout.write('Converting SVG to PNG image ({}/{})       \r'.format(idx, img_count))
				convertToPNG(source, target, S, S)
				if not os.path.isfile(target):
					print('\nERROR: SVG->PNG conversion failed.')
					sys.exit(1)

			converted_count += 1
			if img_exists:
				replaced_count += 1

		# newline after converting files
		print()

		print('\n{} new images added to release ({} updated).'.format(converted_count, replaced_count))
		if removed_count > 0:
			print('{} old images removed from release.'.format(removed_count))

		if live_run:
			copyTemplate(size_dir)
			shutil.copy(file_license, size_dir)
			fileio.write(appendPath(size_dir, 'README.txt'), readme_text)

	if live_run:
		shutil.copy(file_license, dir_release)
		shutil.copy(appendPath(dir_root, 'CHANGES.txt'), dir_release)
		fileio.write(appendPath(dir_release, 'README.txt'), readme_text)

	# create zip distribution archive
	compress(not live_run)

	print('\nDone!')

except KeyboardInterrupt:
	print('\nProcess cancelled by user')
	sys.exit(0)
